### DISCUSSION TITLE
How to deal with DataFrames using too much memory

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#17616] [Q&A] [answered]
Adapted from: https://dagster.slack.com/archives/C01U954MEER/p1698867974627389

I just recently encountered an out-of-memory error. The entire asset size in BigQuery is only 4.5Gb, but it does seem like the conversion to a dataframe is consuming a lot of memory. It seems like the best approach here is to partition this asset, but I don’t fully understand the memory model around partitioned assets, so while I can do this, I want to check that this will solve my problem before I go down the rabbit hole.

For additional context, I’m using the BQ I/O manager, and loading the results of the 4.5Gb query into memory, doing some processing, and then returning it as a single table asset in BQ.

**Comment by sryza** [comment, marked as answer]
Pandas is often not the best framework choice for processing large data. While Dagster has built-in IO managers that load data from Snowflake and BigQuery into Pandas DataFrames, we don't recommend using these if the data that would be loaded is too big to fit in memory.

Dagster is an orchestrator, not a data processing engine, and thus it is largely unopinionated on what engine to use when working with data that does not fit in memory.

### Ideas to help deal with data not fitting in memory

- Try out the suggestions in the Pandas user guide: https://pandas.pydata.org/docs/user_guide/scale.html
- Distribute your computation using a framework like Spark or Dask
- Write a custom IO manager that loads and writes data in a mini-batched or streaming fashion
- Don't use IO managers, and just write your own code inside your `@asset`-decorated function that loads and writes data in a mini-batched or streaming fashion
- Instead of Pandas, use Polars, which can spill to disk
- Instruct the Snowflake/BigQuery IO manager to only load the subset of columns that your computation needs:
    ```python
        @asset(
            ins={"my_table": AssetIn("my_table", metadata={"columns": ["a"]})}
        )
        def my_table_a(my_table: pd.DataFrame) -> pd.DataFrame:
            # my_table will just contain the data from column "a"
            ...
    ```

### What about partitions?

Using Dagster partitions to partition your data will reduce the size of data processed by any particular step.

However, note that Dagster partitions are primarily meant to be an operational and organizational construct, not a parallelism construct. I.e. the main reason to use Dagster partitions is that you want the ability to launch runs that materialize a subset of your asset without touching the rest of it.

If your primarily envision yourself launching runs that process all your data at once, but want that processing to happen in smaller chunks in parallel, then we'd generally recommend using a framework like Spark or Dask to implement this parallelism, instead of Dagster partitions.

**Comment by doctorpangloss** [comment]
What should I do for `Output`s that are too large for memory?

What should I do when the computation and writing happen at the same time, such as authoring SQL statement to be run in Postgres?

I am confused about how `asset`s want concrete, in memory inputs and outputs, which suggests that despite what you've written, things are much more strongly coupled than a `dask` dataframe call can deal with.
